source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+ *   ResponseType = envoy::service::discovery::v3::DiscoveryResponse), and
+ * Delta-xDS (RequestType = envoy::service::discovery::v3::DeltaDiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DeltaDiscoveryResponse).
+ * Both the primary and failover streams are either SotW or Delta-xDS.
+ *
+ * The use of this class will be as follows: the GrpcMux object will own an instance of
+ * the GrpcMuxFailoverManager. The GrpcMuxFailoverManager will own 2 GrpcStreams, primary
+ * and secondary. Each of the primary and secondary streams will invoke GrpcStreamCallbacks
+ * on their corresponding objects (also owned by the GrpcMuxFailoverProxy). These invocations
+ * will be followed by the GrpcMuxFailoverProxy calling the GrpcStreamCallbacks on the GrpcMux
+ * object that initialized it.
+ *
+ * Note: this class is WIP.
+ */
+template <class RequestType, class ResponseType> class GrpcMuxFailover {
+public:
+  // A GrpcStream creator function that receives the stream callbacks and returns a
+  // GrpcStream object. This is introduced to facilitate dependency injection for
+  // testing and will be used to create the primary and failover streams.
+  using GrpcStreamCreator = std::function<GrpcStreamInterfacePtr<RequestType, ResponseType>(
+      GrpcStreamCallbacks<ResponseType>* stream_callbacks)>;
+
+  GrpcMuxFailover(GrpcStreamCreator primary_stream_creator,
+                  OptRef<GrpcStreamCreator> failover_stream_creator,
+                  GrpcStreamCallbacks<ResponseType>& grpc_mux_callbacks)
+      : currently_used_service_(ServiceOptions::None), grpc_mux_callbacks_(grpc_mux_callbacks),
+        primary_callbacks_(*this),
+        primary_grpc_stream_(std::move(primary_stream_creator(&primary_callbacks_))) {
+    ASSERT(primary_grpc_stream_ != nullptr);
+    // At the moment failover isn't implemented.
+    ASSERT(!failover_stream_creator.has_value());
+  }
+
+  virtual ~GrpcMuxFailover() = default;
+
+  void establishNewStream() {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    primary_grpc_stream_->establishNewStream();
+  }
+
+  bool grpcStreamAvailable() const {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    return primary_grpc_stream_->grpcStreamAvailable();
+  }
+
+  void sendMessage(const RequestType& request) {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    primary_grpc_stream_->sendMessage(request);
+  }
+
+  void maybeUpdateQueueSizeStat(uint64_t size) {
+    // TODO(adisuissa): At the moment this is a pass-through method. Once the
+    // implementation matures, this call will be updated.
+    primary_grpc_stream_->maybeUpdateQueueSizeStat(size);
+  }
+
+  bool checkRateLimitAllowsDrain() {
```

Comment:
One thing that I wondered was whether proxying stream calls like this one has clean semantics when the stream changes underneath. E.g. let's say you are checking some availability status and then do some send operation, what happens if there is a TOCTOU issue? 

I imagine we don't end up failing over inside a single call stack, is that the contract with the caller around this kind of consideration?

Reply:
You are correct that some functions will be a no-op/return-false when neither the primary and failover are available.
IMHO the issue is that there is an underlying assumption for some of these methods can only operate if some other condition holds (for example `sendMessage()` should only be used if `grpcStreamAvailable` returns true.

I'm not sure if TOCTOU can happen here, as this is all happening on a single thread, and each call that has this dependency assumes a clear causal dependency between the operations. In other words, I think the same issue could happen without this change, but it doesn't because the user of this class (GrpcMux) assumes causality between the different operations in a series of calls. I think we can probably add more ASSERTs to validate this, if we don't want to modify the interface.

As a side note, you are highlighting a different issue which is `grpcStreamAvailable` doesn't really imply that the stream will be connected to a server, but this is not the actual case. In a future PR this behavior will be updated.

Reply:
Yeah, I think that's fair. I think it should just be clear when failover can occur as part of the contract in comments.